src/Init/System/IO.lean

```diff
@@ -594,6 +586,28 @@ end System.FilePath
 
 namespace IO
 
+namespace FS
+
+def readBinFile (fname : FilePath) : IO ByteArray := do
+  -- Requires metadata so defined after metadata
+  let mdata ← fname.metadata
+  let size := mdata.byteSize.toUSize
+  let handle ← IO.FS.Handle.mk fname .read
+  let buf ←
+    if size > 0 then
+      handle.read mdata.byteSize.toUSize
+    else
+      pure <| ByteArray.mkEmpty 0
+  handle.readBinToEndInto buf
+
+def readFile (fname : FilePath) : IO String := do
+  let data ← readBinFile fname
+  match String.fromUTF8? data with
```

Comment:
Does this make a copy of the data?

Reply:
Yes, unfortunately the runtime representation of `ByteArray` has the data part inlined instead of as a pointer and the layout of the `ByteArray` and `String` object are not compatible :/.